pkg/disk/partition_table.go

```diff
@@ -160,6 +161,282 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
 	return newPT, nil
 }
 
+func mkBIOSBoot() Partition {
+	return Partition{
+		Size:     1 * common.MiB,
+		Bootable: true,
+		Type:     BIOSBootPartitionGUID,
+		UUID:     BIOSBootPartitionUUID,
+	}
+}
+
+func mkESP(size uint64) Partition {
+	return Partition{
+		Size: size,
+		Type: EFISystemPartitionGUID,
+		UUID: EFISystemPartitionUUID,
+		Payload: &Filesystem{
+			Type:         "vfat",
+			UUID:         EFIFilesystemUUID,
+			Mountpoint:   "/boot/efi",
+			Label:        "EFI-SYSTEM",
+			FSTabOptions: "defaults,uid=0,gid=0,umask=077,shortname=winnt",
+			FSTabFreq:    0,
+			FSTabPassNo:  2,
+		},
+	}
+}
+
+type CustomPartitionTableOptions struct {
+	// PartitionTableType must be either "dos" or "gpt". Defaults to "gpt".
+	PartitionTableType string
+
+	// BootMode determines the types of boot-related partitions that are
+	// automatically added, BIOS boot (legacy), ESP (UEFI), or both (hybrid).
+	// If none, no boot-related partitions are created.
+	BootMode platform.BootMode
+
+	// DefaultFSType determines the filesystem type for automatically created
+	// filesystems and custom mountpoints that don't specify a type.
+	// None is only valid if no partitions are created and all mountpoints
+	// partitions specify a type.
+	DefaultFSType FSType
+
+	// RequiredSizes defines a map of minimum sizes for specific directories.
+	// These indirectly control the minimum sizes of partitions. A directory
+	// with a required size will set the minimum size of the partition with the
+	// mountpoint that contains the directory. Additional directory
+	// requirements are additive, meaning the minimum size for a mountpoint's
+	// partition is the sum of all the required directory sizes it will
+	// contain. By default, if no requiredSizes are provided, the new partition
+	// table will require at least 1 GiB for '/' and 2 GiB for '/usr'. In most
+	// cases, this translates to a requirement of 3 GiB for the root partition,
+	// Logical Volume, or Btrfs subvolume.
+	RequiredSizes map[string]uint64
+}
+
+// NewCustomPartitionTable creates a partition table based almost entirely on the partitioning customizations from a blueprint.
+func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization, options *CustomPartitionTableOptions, rng *rand.Rand) (*PartitionTable, error) {
```

Comment:
(nitpick) This is a long function. It's okay because it is very linear so it's not a big deal, but I was wondering if it might be split up so that parts can be reused for otk and also if it would be easier to follow/extend with splitting (and test individually). I did a quick experiment in  https://github.com/osbuild/images/commit/acb38251b703bb3d4564ad2d2f8769ea864ce226#diff-ffc14a35d85eb6fd9724427b51786e40bc3e910bc41cbd6762ab977ae9689765R455 which I kinda like but maybe just personal preferences - can of course be a followup

Reply:
Thanks for extracting helpers here, it looks much nicer now IMHO